### PR TITLE
Log event publishing errors as Err (was Info)

### PR DESCRIPTION
**Pull request by devenney**
Hi folks!

Today I ran into an issue regarding the classification of log messages. Our logging fell over and we saw no errors in the Filebeat logs. After further investigation we found that this error was being logged as INFO:

    2016/08/08 10:56:03.814969 single.go:130: INFO Connecting error publishing events (retrying): Head http://confidential-hostname:9200: dial tcp: lookup confidential-hostname on 10.251.0.2:53: no such host

This feels wrong as any event publishing error seems to be a real error. This PR ensures these errors are logged as errors.

____

We also found the following which I have not modified in this PR as I am not sure if it is intentional behaviour:

    2016/08/08 10:56:20.807266 log.go:80: INFO Read line error: No more bytes

This appears to be an issue in `filebeat/harvester/log.go`, lines 67-85. A number of these logging calls appear to be real error cases which I would argue should be logged as such. If this is the case, I'm happy to update this PR to modify that behaviour too.

### CODE REVIEW

**Comment by ruflin** [comment]
Hi @devenney 

Thanks for opening this PR. We had previously quite a few discussion on this log message if it should be error or info. Some discussions can for example be found here: https://github.com/elastic/beats/issues/875 We decided to go with Info as retrying could happen quite often and only log an error if it completely fails. The issues was kind of improved by setting the default log level to info for the 5.0 releases.

Let me know your thoughts.

**Comment by devenney** [comment]
The problem I see is that info is generally too verbose for production. Perhaps it's personal preference but I only care about the errors, whether they're transient or not. In the case described above it was a real error which would retry forever without success as the DNS record was incorrectly entered. Other machines had Filebeat processes which presumably cached the old DNS so the lack of error logs on fresh machines meant we responded too slow - thankfully in staging but others may not be so lucky. 

Perhaps there's a distinction to be drawn between temporary connection issues and those requiring intervention, but I'd argue that it's impossible to distinguish - this should be the user's responsibility. And honestly, in production, seeing a transient connection problem is quite serious and requires addressing IMO. I see this case was put forward a couple of times in the referenced issue. 

So to summarise, I think this strategy is okay for those who leave logging at the new default level (probably most users) but not great for those who want less verbosity. In our case Filebeat appeared to just hang ad infinitum without any error output so it was the last part of the chain we investigated. I strongly believe this should be printed as Error as it logs real errors regardless of their frequency. 

Slightly off-topic: `filebeat.yml` and `filebeat.full.yml` still say the default log level is error btw - I had no idea the default was info. Unless the level is set in two places? I am not solely responsible for our deployment so it's maybe just a gap in my knowledge.

**Comment by pinkisemils** [comment]
I really want to see this error in my logs next time Filebeat can't find it's way to Elasticsearch.

**Comment by tsg** [comment]
> Slightly off-topic: filebeat.yml and filebeat.full.yml still say the default log level is error btw - I had no idea the default was info.

That's an oversight on my part, I'll fix it.

As part of https://github.com/elastic/beats/issues/1931 we reduced the verbosity of Info and Warn levels down to startup information and only important notices like this one, so there shouldn't be any spamming of the logs.

**Comment by devenney** [comment]
@tsg: Is it an important notice or an important error? The thing is that it feels like a miscategorisation of this log output. I think the changes you referenced are actually a very good step towards having the perfect balance in the logging, but would you not agree that logging an _error_ as _info_ invalidates the levels? I feel that this forces everyone to run the tool with a log level of _info_ as _critical_ and _error_ do not actually output all errors as you would expect. This can be deceptive and trip up users as they see no errors and assume the problem lies elsewhere.

Additionally, @ruffin mentioned "[you] decided to go with Info as retrying could happen quite often" which suggests the original decision was made in the knowledge that the default logging level was _error_. As this is no longer the case it would seem that this particular conclusion of the original discussion is no longer relevant.

Apologies for being _that_ guy who turns a four character PR into a philosophical debate but I have a lot of love for this project and feel that this is quite an important issue which presumably reaches deeper than just this one line of code.

tl;dr: Error logging does not log all errors, therefore users must always run with a logging level of Info - thus the Error level feels moot.

**Comment by tsg** [comment]
@devenney Yeah, I understand your point of view, and I'm also on the fence regarding the log level for that message. While it is an "error", it can be one that is corrected automatically without any need for user intervention. Filebeat is designed to handle transient errors without data loss, so in some cases there is no error from the user point of view. It could be that one Logstash instance was taken out of the rotation or that a Logstash instance didn't start yet. In these cases, an Info message seems more appropriate.

As the person that designed the whole outputs code, I know that @urso also has some opinions over this.

**Comment by urso** [comment]
+1000 for this PR

Please also update the load balancing workers:
- https://github.com/elastic/beats/blob/master/libbeat/outputs/mode/lb/async_worker.go#L79
- https://github.com/elastic/beats/blob/master/libbeat/outputs/mode/lb/sync_worker.go#L79

There have been many changes to outputs since #875, which was used on another related error potentially spamming logs (even on transient failures) masking potential problems in logs. This PR about (re-)connect failures will not be printed per event, but per failed connection-attempt subject to exponential backoff. Plus backoff support was improved in last few months to remember state instead of being reset between attempts. 

If reconnect is required and succeeds immediately, no error message will be printed.

**Comment by tsg** [comment]
Ok, I thought @urso would be against it. In this case, I'm also +1 on the change :-).

**Comment by devenney** [comment]
All done - and somehow caused the build environment gods to smile upon me. 😆 Let me know if there is any further work required on my side.

**Comment by ruflin** [comment]
@devenney Thanks for pushing the discussion. I remember that we also set it to Info as some users were confused by it, but I agree it is better to have one error too much then one too few. +1 on this change.

LGTM
